CorsixTH/Lua/endconditions.lua

```diff
@@ -0,0 +1,254 @@
+      if not tmp_table[crit_name] or tmp_table[crit_name].gap > crit_table.gap then
+        tmp_table[crit_name] = crit_table
+      end
+    end
+  end
+  -- Move into a numbered table
+  for _, crit_table in pairs(tmp_table) do
+    table.insert(report_table, crit_table)
+    count = count + 1
+    if count == 5 then break end
+  end
+
+  -- Fill up the report table with win criteria not already present as lose criteria
+  for i = 1, #local_criteria_variable do
+    local name = local_criteria_variable[i].name
+    if win_group[name] and not tmp_table[name] then
+      count = count + 1
+      report_table[count] = win_group[name]
+      if count == 5 then break end
+    end
+  end
+
+  -- Some criteria icons shouldn't be next to each other
+  table.sort(report_table, function(a,b) return a.criterion < b.criterion end)
+  return report_table
+end
+
+--!param hospital (Hospital) The hospital of the tests.
+--!param lose_table (table) A group of lose conditions from level_config.
+--!param report (boolean) Whether a report table will be returned.
+--!return Losing criteria name and the limit breached,
+-- or if report is true, the report table.
+function EndConditions:checkLoseGroup(hospital, lose_table, report)
+  local report_table, met_count, total_count, reason, limit = {}, 0, 0
+  for crit_name, crit_table in pairs(lose_table) do
+    local target = report and crit_table.boundary or crit_table.lose_value
+    local max_min = crit_table.max_min_lose == 1 and 1 or -1
+    local measure = self:getAttribute(hospital, crit_name)
+    if (measure - target) * max_min > 0 then
+      if report then -- Collect the criteria that should be reported on
+        report_table[crit_name] = crit_table
+        report_table[crit_name].gap = math.abs(measure - target)
+      else
+        reason, limit = crit_name, crit_table.lose_value
+        met_count = met_count + 1
+      end
+    end
+    total_count = total_count + 1
+  end
+  if report then return report_table end
+  -- Have all criteria of the group been met?
+  if met_count == total_count then
+    -- The latest, probably only, criterion met for the lose message
+    return reason, limit
+  end
+end
+
+--!param hospital (Hospital) The hospital of the tests.
+--!param win_table (table) A group of win conditions from level_config.
+--!param findbest (boolean) Whether to return the score of the group.
+--!return string "win" if the group of win conditions is met,
+-- or if findbest is true, the score of the group.
+function EndConditions:checkWinGroup(hospital, win_table, findbest)
+  local met_count, total_count = 0, 0
+  for crit_name, crit_table in pairs(win_table) do
+    local max_min = crit_table.max_min_win == 1 and 1 or -1
+    if (self:getAttribute(hospital, crit_name) - crit_table.win_value) * max_min >= 0 then
+      met_count = met_count + 1
+    end
+    total_count = total_count + 1
+  end
+
+  if findbest then
+    if met_count > 0 and total_count > 0 then
```

Comment:
`total_count` is always at least `met_count`, so if the former holds, there is no need to test the latter.

You can add an assert if you are afraid this relation may get broken.